### PR TITLE
Add functions `base32Encode` and `base32Decode`

### PR DESCRIPTION
**Pull request by cucumber95**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add functions `base32Encode` and `base32Decode`

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 0c04e3bbcf14d96c08aa1d7847c9626f429c91a1 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/63676/0c04e3bbcf14d96c08aa1d7847c9626f429c91a1/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>❌ failure</td></tr>
<tbody>
</table>

File: src/Functions/FunctionBase32Conversion.cpp [hunk 1]
```diff
@@ -0,0 +1,154 @@
+#include <Columns/ColumnConst.h>
```
Comment (rschu1ze) [inline src/Functions/FunctionBase32Conversion.cpp:1]: Let's add SQL tests + in-source docs + markdown docs.

File: src/Functions/FunctionBase32Conversion.cpp [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+namespace
+{
+
+struct Base32Encode
+{
+    static constexpr auto name = "base32Encode";
+
+    static void process(const ColumnString & src_column, ColumnString::MutablePtr & dst_column, size_t input_rows_count)
+    {
+        size_t dst_column_data_size = 0;
```
Comment (rschu1ze) [inline src/Functions/FunctionBase32Conversion.cpp:32]: My attempt to run `select base32Encode('Hello World')` crashed the server:

```
[ip-172-31-2-215] 2024.05.19 20:01:35.903395 [ 580047 ] <Fatal> BaseDaemon: ########################################
[ip-172-31-2-215] 2024.05.19 20:01:35.903601 [ 580047 ] <Fatal> BaseDaemon: (version 24.5.1.1, build id: 02F01DD9E8443D21782B484068BA9B3D90D6CB3C, git hash: 0f77341ad3600c488d335fc4ffa83a9bb4e984b1) (from thread 577968) (query_id: 9e2a5a49-34e8-475b-9e80-83c50814965e) (query: select base32Encode('Hello World')) Received signal Aborted (6)
[ip-172-31-2-215] 2024.05.19 20:01:35.903684 [ 580047 ] <Fatal> BaseDaemon:
[ip-172-31-2-215] 2024.05.19 20:01:35.903773 [ 580047 ] <Fatal> BaseDaemon: Stack trace: 0x000000000ca209ca 0x000000000ca208c0 0x000000000cdf5545 0x00007afd95c45320 0x00007afd95c9eb1d 0x00007afd95c4526e 0x00007afd95c288ff 0x00007afd95c2881b 0x00007afd95c3b507 0x000000000ca044dd 0x0000000004d9c932 0x0000000004d9c341 0x0000000004d8bd59 0x00000000053e840d 0x0000000012219126 0x0000000012218ca3 0x0000000012219003 0x00000000
1221a246 0x000000001221b1b5 0x00000000133f7930 0x00000000133dd08a 0x00000000133db501 0x00000000133e4ea7 0x00000000133d6ea3 0x00000000133d4fb4 0x00000000133d4c73 0x00000000133c5726 0x000000001412662c 0x000000001412524e 0x00000000141283b7 0x0000000014127acc 0x0000000014127a7d 0x0000000014127a1d 0x00000000141279dd 0x00000000141279a5 0x0000000013769d3e 0x00000000137652bd 0x000000001376494d 0x0000000014779c50 0x000000001477
4bea 0x0000000015c962b9 0x0000000015caae3f 0x00000000170d6c79 0x00000000170d74c8 0x0000000017168c94
[ip-172-31-2-215] 2024.05.19 20:01:35.921322 [ 580047 ] <Fatal> BaseDaemon: 0. /repo/ch1/src/Common/StackTrace.cpp:349: StackTrace::tryCapture() @ 0x000000000ca209ca
[ip-172-31-2-215] 2024.05.19 20:01:35.936249 [ 580047 ] <Fatal> BaseDaemon: 1. /repo/ch1/src/Common/StackTrace.cpp:323: StackTrace::StackTrace(ucontext_t const&) @ 0x000000000ca208c0
[ip-172-31-2-215] 2024.05.19 20:01:35.965660 [ 580047 ] <Fatal> BaseDaemon: 2. /repo/ch1/src/Daemon/BaseDaemon.cpp:154: signalHandler(int, siginfo_t*, void*) @ 0x000000000cdf5545
[ip-172-31-2-215] 2024.05.19 20:01:35.965778 [ 580047 ] <Fatal> BaseDaemon: 3. ? @ 0x00007afd95c45320
[ip-172-31-2-215] 2024.05.19 20:01:35.965898 [ 580047 ] <Fatal> BaseDaemon: 4. ? @ 0x00007afd95c9eb1d
[ip-172-31-2-215] 2024.05.19 20:01:35.965990 [ 580047 ] <Fatal> BaseDaemon: 5. ? @ 0x00007afd95c4526e
[ip-172-31-2-215] 2024.05.19 20:01:35.966088 [ 580047 ] <Fatal> BaseDaemon: 6. ? @ 0x00007afd95c288ff
[ip-172-31-2-215] 2024.05.19 20:01:35.966183 [ 580047 ] <Fatal> BaseDaemon: 7. ? @ 0x00007afd95c2881b
[ip-172-31-2-215] 2024.05.19 20:01:35.966275 [ 580047 ] <Fatal> BaseDaemon: 8. ? @ 0x00007afd95c3b507
[ip-172-31-2-215] 2024.05.19 20:01:35.974661 [ 580047 ] <Fatal> BaseDaemon: 9. /repo/ch1/src/Common/PODArray.h:0: DB::PODArray<char8_t, 4096ul, Allocator<false, false>, 63ul, 64ul>::operator[](long) @ 0x000000000ca044dd
[ip-172-31-2-215] 2024.05.19 20:01:35.985722 [ 580047 ] <Fatal> BaseDaemon: 10. /repo/ch1/src/Functions/FunctionBase32Conversion.cpp:56: DB::(anonymous namespace)::Base32Encode::process(DB::ColumnString const&, COW<DB::IColumn>::mutable_ptr<DB::ColumnString>&, unsigned long) @ 0x0000000004d9c932
[ip-172-31-2-215] 2024.05.19 20:01:35.996666 [ 580047 ] <Fatal> BaseDaemon: 11. /repo/ch1/src/Functions/FunctionBase32Conversion.cpp:140: DB::(anonymous namespace)::FunctionBase32Conversion<DB::(anonymous namespace)::Base32Encode>::executeImpl(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long) const @ 0x0000000004d9c341
[ip-172-31-2-215] 2024.05.19 20:01:36.009574 [ 580047 ] <Fatal> BaseDaemon: 12. /repo/ch1/src/Functions/IFunction.h:432: DB::IFunction::executeImplDryRun(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long) const @ 0x0000000004d8bd59
[ip-172-31-2-215] 2024.05.19 20:01:36.189522 [ 580047 ] <Fatal> BaseDaemon: 13. /repo/ch1/src/Functions/IFunctionAdaptors.h:26: DB::FunctionToExecutableFunctionAdaptor::executeDryRunImpl(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long) const @ 0x00000000053e840d
[ip-172-31-2-215] 2024.05.19 20:01:36.203528 [ 580047 ] <Fatal> BaseDaemon: 14. /repo/ch1/src/Functions/IFunction.cpp:246: DB::IExecutableFunction::executeWithoutLowCardinalityColumns(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x0000000012219126
[ip-172-31-2-215] 2024.05.19 20:01:36.216394 [ 580047 ] <Fatal> BaseDaemon: 15. /repo/ch1/src/Functions/IFunction.cpp:160: DB::IExecutableFunction::defaultImplementationForConstantArguments(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x0000000012218ca3
[ip-172-31-2-215] 2024.05.19 20:01:36.229252 [ 580047 ] <Fatal> BaseDaemon: 16. /repo/ch1/src/Functions/IFunction.cpp:238: DB::IExecutableFunction::executeWithoutLowCardinalityColumns(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x0000000012219003
[ip-172-31-2-215] 2024.05.19 20:01:36.242217 [ 580047 ] <Fatal> BaseDaemon: 17. /repo/ch1/src/Functions/IFunction.cpp:303: DB::IExecutableFunction::executeWithoutSparseColumns(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x000000001221a246
[ip-172-31-2-215] 2024.05.19 20:01:36.255226 [ 580047 ] <Fatal> BaseDaemon: 18. /repo/ch1/src/Functions/IFunction.cpp:378: DB::IExecutableFunction::execute(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x000000001221b1b5
[ip-172-31-2-215] 2024.05.19 20:01:36.357915 [ 580047 ] <Fatal> BaseDaemon: 19. /repo/ch1/src/Analyzer/Passes/QueryAnalysisPass.cpp:6163: DB::(anonymous namespace)::QueryAnalyzer::resolveFunction(std::shared_ptr<DB::IQueryTreeNode>&, DB::(anonymous namespace)::IdentifierResolveScope&) @ 0x00000000133f7930
[ip-172-31-2-215] 2024.05.19 20:01:36.459160 [ 580047 ] <Fatal> BaseDaemon: 20. /repo/ch1/src/Analyzer/Passes/QueryAnalysisPass.cpp:6470: DB::(anonymous namespace)::QueryAnalyzer::resolveExpressionNode(std::shared_ptr<DB::IQueryTreeNode>&, DB::(anonymous namespace)::IdentifierResolveScope&, bool, bool) @ 0x00000000133dd08a
[ip-172-31-2-215] 2024.05.19 20:01:36.560285 [ 580047 ] <Fatal> BaseDaemon: 21. /repo/ch1/src/Analyzer/Passes/QueryAnalysisPass.cpp:6622: DB::(anonymous namespace)::QueryAnalyzer::resolveExpressionNodeList(std::shared_ptr<DB::IQueryTreeNode>&, DB::(anonymous namespace)::IdentifierResolveScope&, bool, bool) @ 0x00000000133db501
[ip-172-31-2-215] 2024.05.19 20:01:36.663089 [ 580047 ] <Fatal> BaseDaemon: 22. /repo/ch1/src/Analyzer/Passes/QueryAnalysisPass.cpp:6879: DB::(anonymous namespace)::QueryAnalyzer::resolveProjectionExpressionNodeList(std::shared_ptr<DB::IQueryTreeNode>&, DB::(anonymous namespace)::IdentifierResolveScope&) @ 0x00000000133e4ea7
[ip-172-31-2-215] 2024.05.19 20:01:36.763889 [ 580047 ] <Fatal> BaseDaemon: 23. /repo/ch1/src/Analyzer/Passes/QueryAnalysisPass.cpp:8119: DB::(anonymous namespace)::QueryAnalyzer::resolveQuery(std::shared_ptr<DB::IQueryTreeNode> const&, DB::(anonymous namespace)::IdentifierResolveScope&) @ 0x00000000133d6ea3
[ip-172-31-2-215] 2024.05.19 20:01:36.864105 [ 580047 ] <Fatal> BaseDaemon: 24. /repo/ch1/src/Analyzer/Passes/QueryAnalysisPass.cpp:1158: DB::(anonymous namespace)::QueryAnalyzer::resolve(std::shared_ptr<DB::IQueryTreeNode>&, std::shared_ptr<DB::IQueryTreeNode> const&, std::shared_ptr<DB::Context const>) @ 0x00000000133d4fb4
[ip-172-31-2-215] 2024.05.19 20:01:36.964380 [ 580047 ] <Fatal> BaseDaemon: 25. /repo/ch1/src/Analyzer/Passes/QueryAnalysisPass.cpp:8414: DB::QueryAnalysisPass::run(std::shared_ptr<DB::IQueryTreeNode>&, std::shared_ptr<DB::Context const>) @ 0x00000000133d4c73
[ip-172-31-2-215] 2024.05.19 20:01:36.986621 [ 580047 ] <Fatal> BaseDaemon: 26. /repo/ch1/src/Analyzer/QueryTreePassManager.cpp:185: DB::QueryTreePassManager::run(std::shared_ptr<DB::IQueryTreeNode>) @ 0x00000000133c5726
[ip-172-31-2-215] 2024.05.19 20:01:37.018189 [ 580047 ] <Fatal> BaseDaemon: 27. /repo/ch1/src/Interpreters/InterpreterSelectQueryAnalyzer.cpp:142: DB::(anonymous namespace)::buildQueryTreeAndRunPasses(std::shared_ptr<DB::IAST> const&, DB::SelectQueryOptions const&, std::shared_ptr<DB::Context const> const&, std::shared_ptr<DB::IStorage> const&) @ 0x000000001412662c
[ip-172-31-2-215] 2024.05.19 20:01:37.045997 [ 580047 ] <Fatal> BaseDaemon: 28. /repo/ch1/src/Interpreters/InterpreterSelectQueryAnalyzer.cpp:160: DB::InterpreterSelectQueryAnalyzer::InterpreterSelectQueryAnalyzer(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const> const&, DB::SelectQueryOptions const&, std::vector<String, std::allocator<String>> const&) @ 0x000000001412524e
[ip-172-31-2-215] 2024.05.19 20:01:37.077297 [ 580047 ] <Fatal> BaseDaemon: 29. /repo/ch1/contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:714: std::__unique_if<DB::InterpreterSelectQueryAnalyzer>::__unique_single std::make_unique[abi:v15000]<DB::InterpreterSelectQueryAnalyzer, std::shared_ptr<DB::IAST>&, std::shared_ptr<DB::Context> const&, DB::SelectQueryOptions const&>(std::shared_ptr<DB::IAST>&, std::share
d_ptr<DB::Context> const&, DB::SelectQueryOptions const&) @ 0x00000000141283b7
[ip-172-31-2-215] 2024.05.19 20:01:37.107760 [ 580047 ] <Fatal> BaseDaemon: 30. /repo/ch1/src/Interpreters/InterpreterSelectQueryAnalyzer.cpp:270: DB::registerInterpreterSelectQueryAnalyzer(DB::InterpreterFactory&)::$_0::operator()(DB::InterpreterFactory::Arguments const&) const @ 0x0000000014127acc
[ip-172-31-2-215] 2024.05.19 20:01:37.138831 [ 580047 ] <Fatal> BaseDaemon: 31. /repo/ch1/contrib/llvm-project/libcxx/include/__functional/invoke.h:394: ? @ 0x0000000014127a7d
[ip-172-31-2-215] 2024.05.19 20:01:37.169365 [ 580047 ] <Fatal> BaseDaemon: 32. /repo/ch1/contrib/llvm-project/libcxx/include/__functional/invoke.h:470: ? @ 0x0000000014127a1d
[ip-172-31-2-215] 2024.05.19 20:01:37.199798 [ 580047 ] <Fatal> BaseDaemon: 33. /repo/ch1/contrib/llvm-project/libcxx/include/__functional/function.h:235: ? @ 0x00000000141279dd
[ip-172-31-2-215] 2024.05.19 20:01:37.230444 [ 580047 ] <Fatal> BaseDaemon: 34. /repo/ch1/contrib/llvm-project/libcxx/include/__functional/function.h:716: ? @ 0x00000000141279a5
[ip-172-31-2-215] 2024.05.19 20:01:37.247818 [ 580047 ] <Fatal> BaseDaemon: 35. /repo/ch1/contrib/llvm-project/libcxx/include/__functional/function.h:848: ? @ 0x0000000013769d3e
[ip-172-31-2-215] 2024.05.19 20:01:37.263106 [ 580047 ] <Fatal> BaseDaemon: 36. /repo/ch1/contrib/llvm-project/libcxx/include/__functional/function.h:1187: ? @ 0x00000000137652bd
[ip-172-31-2-215] 2024.05.19 20:01:37.278292 [ 580047 ] <Fatal> BaseDaemon: 37. /repo/ch1/src/Interpreters/InterpreterFactory.cpp:355: DB::InterpreterFactory::get(std::shared_ptr<DB::IAST>&, std::shared_ptr<DB::Context>, DB::SelectQueryOptions const&) @ 0x000000001376494d
[ip-172-31-2-215] 2024.05.19 20:01:37.334003 [ 580047 ] <Fatal> BaseDaemon: 38. /repo/ch1/src/Interpreters/executeQuery.cpp:1138: DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x0000000014779c50
[ip-172-31-2-215] 2024.05.19 20:01:37.390821 [ 580047 ] <Fatal> BaseDaemon: 39. /repo/ch1/src/Interpreters/executeQuery.cpp:1395: DB::executeQuery(String const&, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum) @ 0x0000000014774bea
[ip-172-31-2-215] 2024.05.19 20:01:37.450307 [ 580047 ] <Fatal> BaseDaemon: 40. /repo/ch1/src/Server/TCPHandler.cpp:522: DB::TCPHandler::runImpl() @ 0x0000000015c962b9
[ip-172-31-2-215] 2024.05.19 20:01:37.511507 [ 580047 ] <Fatal> BaseDaemon: 41. /repo/ch1/src/Server/TCPHandler.cpp:2341: DB::TCPHandler::run() @ 0x0000000015caae3f
```

File: src/Common/Base32.cpp [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    {
+        if (elem >= 'A' && elem <= 'Z')
+            return elem - 'A';
+        else if (elem >= '2' && elem <= '7')
+            return elem - '2' + 26;
+        return -1;
+    }
+}
+
+void encodeBase32(const UInt8 * src, size_t src_length, UInt8 * dst)
```
Comment (rschu1ze) [inline src/Common/Base32.cpp:19]: I appreciate that you took the time to write your base32 implementation.

The thing is that ClickHouse is not in the base32-implementation business (it is also not in the business of writing unicode processing libraries, regexp libraries, etc.). In fact, base64 encoding/decoding (src/Functions/FunctionBase64Conversion.h) calls out into this [library](https://github.com/aklomp/base64). Yes, base58 is hand-written but I think this was a mistake. For base32, there is a [library](https://github.com/tplgy/cppcodec) with
- acceptable license (MIT),
- header-only build (if you like to integrate it into the Clickhouse build, look at contrib/annoy/ and contrib/annoy-cmake/ which is another header-only library),
- decent API as far as I see
- it supports all of the different base32 alphabet [craziness](https://en.wikipedia.org/wiki/Base32) - not that we would need it but it is nice-to-have anyways.

So my preference would be to proceed similar to base64, i.e. call out to a library instead.

**Review by rschu1ze — APPROVED** [review]

**Review by rschu1ze — CHANGES_REQUESTED** [review]
Hi, thanks for the PR.

**Comment by rschu1ze** [comment]
@cucumber95 Please re-open if you like to continue, thanks.

**Pull request closed without merging**